### PR TITLE
Fixed #29480 -- Made MySQL backend retrieve constraint columns in their defined order.

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

**Comment by adamchainz** [comment]
Does this test fail beforehand, or is it random? Also could you file a ticket with an explanation why this is needed?

**Comment by graingert** [comment]
@adamchainz I can't get the tests to fail. The problem we have is intermittent and I suspect it only happens after multiple migrations add and remove constraints from a table

**Comment by adamchainz** [comment]
Sure, I understand this could be hard to reproduce, especially with the underlying data dictionary changes between 5.6 - 8.0. But can you at least make a ticket so it's not just a random whatsapp message you sent me?

**Comment by graingert** [comment]
Random WhatsApp message for reference:


[14/05, 13:37] Thomas Grainger: @⁨Adam Johnson⁩ does information_schema.key_column_usage have a default ordering?
[14/05, 13:37] Thomas Grainger: someone added this patch to our local django but doesn't know why they did it anymore: https://github.com/django/django/pull/9941/files
[14/05, 13:38] Thomas Grainger: it's something to do with the introspection getting columns back in the wrong order
[14/05, 13:38] Thomas Grainger: and comparing differently when generating migrations
[14/05, 13:39] Adam Johnson: Probably not
[14/05, 13:39] Adam Johnson: Explicit order by is always a good idea
[14/05, 13:39] Thomas Grainger: so it's surprising that they are always introspected in the correct order
[14/05, 13:40] Thomas Grainger: I've added a test but it still passes on Jenkins without my change
[14/05, 13:40] Adam Johnson: Probably there’s an unguaranteed ordering but it could be breakable by doing lots of alters
[14/05, 13:42] Thomas Grainger: hmm, any idea on how to make a test case?
[14/05, 13:42] Thomas Grainger: (and therefore get this merged)

**Comment by adamchainz** [comment]
So `get_constraints` returns a dict, so its order isn't guaranteed until Python 3.7 anyway.

I looked around for uses of `get_constraints` inside Django, and think this is the real patch - the boundary where the dict is turned into a list...

```patch
diff --git a/django/db/backends/base/schema.py b/django/db/backends/base/schema.py
index cf18e7653e..7d99495701 100644
--- a/django/db/backends/base/schema.py
+++ b/django/db/backends/base/schema.py
@@ -1020,7 +1020,7 @@ class BaseDatabaseSchemaEditor:
         with self.connection.cursor() as cursor:
             constraints = self.connection.introspection.get_constraints(cursor, model._meta.db_table)
         result = []
-        for name, infodict in constraints.items():
+        for name, infodict in sorted(constraints.items()):
             if column_names is None or column_names == infodict['columns']:
                 if unique is not None and infodict['unique'] != unique:
                     continue
```

**Comment by graingert** [comment]
We don't have a problem with the order of the constraints themselves just the columns in those constraints: https://github.com/graingert/django/blob/43cc0fb511f989e1eb15137624b72c1e21a0cdd9/django/db/backends/mysql/introspection.py#L169

**Comment by adamchainz** [comment]
Ah right

@MarkusH what do you think about merging this without a test?

**Comment by MarkusH** [comment]
I'd like to have a ticket with a traceback etc. If it's unfeasible to provide a reliable test I'd be fine w/o one. But in that case the ticket should make this clear.

**Comment by graingert** [comment]
@MarkusH https://code.djangoproject.com/ticket/29480#ticket

**Comment by graingert** [comment]
Ok will do that

File: django/db/backends/mysql/introspection.py [hunk 1]
```diff
@@ -160,6 +160,7 @@ def get_constraints(self, cursor, table_name):
             WHERE
                 kc.table_schema = DATABASE() AND
                 kc.table_name = %s
+            ORDER BY kc.`ordinal_position` ASC
```
Comment (felixxm) [inline django/db/backends/mysql/introspection.py]: I think that `ASC` is unnecessary.

Comment (graingert) [reply django/db/backends/mysql/introspection.py]: ASC removed

**Comment by graingert** [comment]
@adamchainz tests removed

**Comment by graingert** [comment]
Can we get a Django 1.11 backport?

**Comment by timgraham** [comment]
As far as I can tell, this doesn't qualify for a backport based on our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions).

**Comment by adamchainz** [comment]
I don't think so either, given the exact bug hasn't been specified.

**Comment by j0hnsmith** [comment]
To hopefully help someone out in the future...

Given a mysql table with a constraint such as

```
UNIQUE KEY `some_model_fk_252ff8b0_uniq` (`c`,`b`,`a`),
```
with the columns in the index specified in the order of the `unique_together` declaration.

Without the `ORDER BY kc.ordinal_position` added in this PR, the query would get the columns (sometimes, but not always???) out of order, eg `a, b, c`... which doesn't match the order `django/django/db/backends/base/schema.py:BaseDatabaseSchemaEditor._constraint_names()` has.

This PR fixes at least one of the reasons for

```
ValueError: Found wrong number (0) of constraints for ...
```

**Comment by charettes** [comment]
> (sometimes, but not always???)

I assume the non-determinism of the issue originates from the fact MySQL (and other SQL implementation) can choose to return columns in whatever order it wants if no `ORDER BY` is specified so they might sometime be in the right order and sometimes not.

**Pull request merged**